### PR TITLE
Feature/csi 4361 resolve space efficiency flashcopy 2 0

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: controller/scripts/lint.ini [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (oriyarde) [inline controller/scripts/lint.ini:1]: ?

File: controller/scripts/lint.ini [hunk 2]
```diff
@@ -44,6 +45,7 @@ disable=
 [BASIC]
 include-naming-hint=y # show enforced regex
 good-names=i,e
+max-branches=13
```
Comment (oriyarde) [inline controller/scripts/lint.ini:48]: do we have a ticket to reduce it?

File: controller/array_action/array_mediator_interface.py [hunk 3]
```diff
@@ -364,6 +364,24 @@ def validate_supported_space_efficiency(self, space_efficiency):
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def validate_space_efficiency_match_source(self, space_efficiency, source_id, source_type):
+        """
+        This function will check if the space efficiency passed to CreateVolume is match the volume source.
+
+        Args:
+            space_efficiency : as passed from the storage class
```
Comment (oriyarde) [inline controller/array_action/array_mediator_interface.py:373]:
```suggestion
            space_efficiency : as passed from the CSI request
```

File: controller/scripts/lint.ini (hunk 2 above)
Comment (ArbelNathan) [reply controller/scripts/lint.ini:48]: CSI-4378

File: controller/scripts/lint.ini (hunk 1 above)
Comment (ArbelNathan) [reply controller/scripts/lint.ini]: removed - done

File: controller/array_action/array_mediator_interface.py (hunk 3 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_interface.py]: done

File: controller/array_action/array_mediator_ds8k.py [hunk 4]
```diff
@@ -697,3 +697,6 @@ def promote_replication_volume(self, replication_name):
 
     def demote_replication_volume(self, replication_name):
         raise NotImplementedError
+
+    def validate_space_efficiency_match_source(self, space_efficiency, source_id, source_type):
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:701]:
```suggestion
    def validate_space_efficiency_matches_source(self, space_efficiency, source_id, source_type):
```

File: controller/array_action/array_mediator_svc.py [hunk 5]
```diff
@@ -1499,3 +1504,15 @@ def _is_vdisk_has_fcmaps(self, vdisk_uid):
...
+    def validate_space_efficiency_match_source(self, space_efficiency, source_id, source_type):
+        if source_type == controller_config.SNAPSHOT_TYPE_NAME:
+            cli_snapshot = self._get_cli_snapshot_by_id(source_id)
+            source_cli_volume = self._get_cli_volume(cli_snapshot.volume_id)
+        else:
+            source_volume_name = self._get_volume_name_by_wwn(source_id)
+            source_cli_volume = self._get_cli_volume(source_volume_name)
+        source_volume_space_efficiency = _get_cli_volume_space_efficiency(source_cli_volume)
+        if not _is_parameter_space_efficiency_match_array_space_efficiency(space_efficiency,
+                                                                           source_volume_space_efficiency):
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:1517]:
```suggestion
        if not _is_requested_space_efficiency_matches_source_space_efficiency(space_efficiency,
                                                                              source_volume_space_efficiency):
```

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:1517]: or more shortly:
```suggestion
        if not _is_requested_space_efficiency_matches_source(space_efficiency,
                                                             source_volume_space_efficiency):
```

File: controller/tests/controller_server/csi_controller_server_test.py [hunk 6]
```diff
@@ -200,6 +200,14 @@ def test_create_snapshot_with_space_efficiency_parameter_succeeds(self, storage_
         self.request.parameters = {config.PARAMETERS_SPACE_EFFICIENCY: SPACE_EFFICIENCY}
         self._test_create_snapshot_succeeds(storage_agent, expected_space_efficiency=SPACE_EFFICIENCY)
 
+    def test_create_snapshot_with_space_efficiency_and_virt_snap_func_enabled_failed(self):
```
Comment (oriyarde) [inline controller/tests/controller_server/csi_controller_server_test.py:203]:
```suggestion
    def test_create_snapshot_with_space_efficiency_and_virt_snap_func_enabled_fail(self):

```

File: controller/array_action/array_mediator_interface.py [hunk 7]
```diff
@@ -364,6 +364,24 @@ def validate_supported_space_efficiency(self, space_efficiency):
...
+        Args:
+            space_efficiency : as passed from the CSI request
+            source_id        : id of source to create from
+            source_type      : volume or snapshot
+
+        Returns:
+            None
+
+        Raises:
+            TargetSpaceEfficiencyMismatch
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_interface.py:381]: done

File: controller/array_action/array_mediator_ds8k.py (hunk 4 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_ds8k.py]: done

File: controller/array_action/array_mediator_svc.py (hunk 5 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py]: done

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 6 above)
Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py]: done

File: controller/array_action/array_mediator_interface.py [hunk 8]
```diff
@@ -364,6 +364,24 @@ def validate_supported_space_efficiency(self, space_efficiency):
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def validate_space_efficiency_match_source(self, space_efficiency, source_id, source_type):
+        """
+        This function will check if the space efficiency passed to CreateVolume is match the volume source.
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_interface.py]: done

File: controller/array_action/array_mediator_interface.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_interface.py:381]: I don't see a change

[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_interface.py]: now

File: controller/array_action/messages.py [hunk 9]
```diff
@@ -14,6 +14,9 @@
 
 SPACE_EFFICIENCY_NOT_SUPPORTED_MESSAGE = "space efficiency is not supported : {0} "
 
+SPACE_EFFICIENCY_MISMATCH_MESSAGE = "space efficiency : {0}" \
+                                           " does not match the source volume space efficiency : {1}"
```
Comment (oriyarde) [inline controller/array_action/messages.py:18]: is this the automatic indentation?

Comment (ArbelNathan) [reply controller/array_action/messages.py]: nope
done

**Review by oriyarde — APPROVED** [review]
LGTM

**Review by matancarmeli7 — APPROVED** [review]
LGTM

**Review by matancarmeli7 — APPROVED** [review]

**Review by oriyarde — APPROVED** [review]
LGTM

**Pull request merged**